### PR TITLE
feat: add network topology functions

### PR DESCRIPTION
**Pull request by fcgxz2003**
Construct the network topology by storing probes in redis.
There are three types of storage structures in redis:

network-topology:src:dest (hash)
Store averageRTT, createdAt, updatedAt of multiple probes from src to dest.
probes:src:dest (list)
Store probes list from src to dest.
visit-times:src (string)
Store visit times of src in this network topology, in other word, the number of times src is probed.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: scheduler/networktopology/network_topology.go [hunk 1]
```diff
@@ -14,33 +14,78 @@
...
-	config *config.Config
-
 	// Redis universal client interface.
 	rdb redis.UniversalClient
 
-	// Resource interface.
+	// Scheduler config.
+	config *config.Config
+
+	// Resource interface
```
Comment (MinH-09) [inline scheduler/networktopology/network_topology.go:85]: Resource interface.

Comment (MIchaelFU0403) [reply scheduler/networktopology/network_topology.go:85]: add period.

File: scheduler/networktopology/network_topology.go [hunk 2]
```diff
@@ -56,5 +101,232 @@ func NewNetworkTopology(cfg *config.Config, rdb redis.UniversalClient, resource
...
+
+	averageRTT, err := time.ParseDuration(value)
+	if err != nil {
+		return time.Duration(0)
+	}
+
+	return averageRTT
+}
+
+// VisitTimes is the visit times of host.
```
Comment (fcgxz2003) [inline scheduler/networktopology/network_topology.go:211]: VisitTimes is the number of times the host has been probed.

File: scheduler/networktopology/network_topology.go [hunk 3]
```diff
@@ -56,5 +101,232 @@ func NewNetworkTopology(cfg *config.Config, rdb redis.UniversalClient, resource
 
 // Peek returns the oldest probe without removing it.
 func (n *networkTopology) Peek(src, dest string) (*Probe, bool) {
-	return nil, false
+	key := fmt.Sprintf("probes:%s:%s", src, dest)
+	jsonStr, err := n.rdb.LIndex(context.Background(), key, 0).Result()
```
Comment (gaius-qi) [inline scheduler/networktopology/network_topology.go:105]: What is `jsonStr`?

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: scheduler/networktopology/network_topology.go [hunk 4]
```diff
@@ -56,5 +101,232 @@ func NewNetworkTopology(cfg *config.Config, rdb redis.UniversalClient, resource
...
+
+// UpdatedAt is the updated time to store probe.
+func (n *networkTopology) UpdatedAt(src, dest string) time.Time {
+	key := fmt.Sprintf("network-topology:%s:%s", src, dest)
+	value, err := n.rdb.HGet(context.Background(), key, "updatedAt").Result()
+	if err != nil {
+		return time.Time{}
+	}
+
+	updatedAt, err := time.Parse(TimeFormat, value)
```
Comment (gaius-qi) [inline scheduler/networktopology/network_topology.go:187]: Use `UnixNano()`.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: scheduler/networktopology/network_topology.go (hunk 3 above)
Comment (fcgxz2003) [reply scheduler/networktopology/network_topology.go]: Because Probe stores in redis as a string parsed into json, called jsonStr.

**Review by gaius-qi — APPROVED** [review]
LGTM

**Pull request merged**